R/utilities-mass-balance.R

```diff
@@ -357,3 +318,162 @@ getMassBalanceDataMapping <- function(plotType) {
+#' @import ospsuite.utils
+#' @import dplyr
+#' @keywords internal
+getApplicationResults <- function(applications, infusionSteps = 20) {
+  applicationResults <- data.frame()
+  for (application in applications) {
+    # No Infusion
+    if (is.null(application$infusionTime$value)) {
+      applicationResults <- rbind.data.frame(
+        applicationResults,
+        data.frame(
+          time = application$startTime$value,
+          drugMass = application$drugMass$value
+        )
+      )
+      next
+    }
+    # For infusion, infusion duration needs to be discretized in multiple time steps
+    applicationResults <- rbind.data.frame(
+      applicationResults,
+      data.frame(
+        time = seq(
+          application$startTime$value,
+          application$startTime$value + application$infusionTime$value,
+          length.out = infusionSteps
+        ),
+        drugMass = application$drugMass$value / infusionSteps
+      )
+    )
+  }
+  # Normalization uses total drug mass
+  applicationResults <- applicationResults %>%
+    mutate(totalDrugMass = cumsum(drugMass))
+  return(applicationResults)
+}
+
+#' @title getMassBalanceData
+#' @description Get mass balance data for a set of compounds
+#' @param groupings A list of grouping lists that define naming and inclusion/exclusion criteria
+#' @param compoundNames A vector of compound names
+#' @param simulation A `Simulation` object
+#' @param simulationResults A `SimulationResults` object
+#' @return A data.frame that includes `Time`, `Amount` and `Legend` as variables
+#' @import ospsuite
+#' @import ospsuite.utils
+#' @keywords internal
+getMassBalanceData <- function(groupings, compoundNames, simulation, simulationResults) {
+  # If groupings is NULL, use default inclusion/exclusion and grouping
+  groupings <- groupings %||% defaultMassBalanceGroupings(simulationResults, compoundNames)
+
+  massBalanceData <- data.frame()
+  previouslyIncludedMoleculePaths <- NULL
+  for (group in groupings) {
+    includedMolecules <- ospsuite::getAllMoleculesMatching(group$Include, simulation)
+    includedMoleculePaths <- sapply(includedMolecules, function(molecule) {
+      molecule$path
+    })
+    # Exclusion criteria
+    excludedMoleculePaths <- NULL
+    # If ExcludePreviousGroupings is TRUE, exclude molecules from previous groupings
+    if(group$ExcludePreviousGroupings %||% TRUE){
+      excludedMoleculePaths <- previouslyIncludedMoleculePaths
+    }
+    if (!is.null(group$Exclude)) {
+      excludedMolecules <- ospsuite::getAllMoleculesMatching(group$Exclude, simulation)
+      excludedMoleculePaths <- c(
+        excludedMoleculePaths,
+        sapply(excludedMolecules, function(molecule){molecule$path})
+      )
+    }
+    includedMolecules <- includedMolecules[!includedMoleculePaths %in% excludedMoleculePaths]
+    includedMoleculePaths <- setdiff(includedMoleculePaths, excludedMoleculePaths)
+    
+    if(isEmpty(includedMoleculePaths)) {
+      next
+    }
```

Comment:
here we should produce a warning that there is nothing to display in the group (might be a user misconfiguration)